crdt/sequence.py

```diff
@@ -0,0 +1,247 @@
+        """
+        Compares this Operation to another Operation to determine ordering.
+        """
+        return a.owner.compare(b.owner)
+
+    def index_transform(self, position):
+        """
+        Returns the transformed index, which is the actual index of the object in the
+        sequence, taking into account tombstones.
+        """
+        if position < 0 or position >= len(self.sequence):
+            raise IndexError("Index out of bounds")
+
+        index = -1
+        while position >= 0:
+            index += 1
+            if not self.sequence[index].tombstone:
+                position -= 1
+        return index
+
+    def append(self, object):
+        """
+        Appends an item to the end of the sequence.
+        """
+        
+        # Tick the clock
+        self.clock.add(1)
+
+        # Add the insert operation to the log and update the sequence
+        op = Operation(OpId(self.id, self.clock.get()), OperationType.INSERT, None, payload=object)
+        self.operations.add(op)
+        op.do(self.sequence)
+
+    def insert(self, position, object):
+        """
+        Inserts an item to the sequence at the specified position. The position is a
+        0-based index from the perspective of the caller (e.g., does not count deleted
+        objects). This makes it easier for the caller to make insertions because the
+        caller does not have to keep track of object IDs.
+        """
+
+        # Tick the clock
+        self.clock.add(1)
+
+        index = self.index_transform(position)
+        target = self.sequence[index].operation
+        op_id = self.clock.get()
+
+        # Add the insert operation to the log and update the sequence
+        op = Operation(OpId(self.id, op_id), OperationType.INSERT, target, payload=object)
+        self.operations.add(op)
+        op.do(self.sequence)
+
+    def remove(self, position):
+        """
+        Removes an item from the set at the specified position. The position must be
+        within bounds of the current sequence, or this raises an IndexError.
+        """
+        # Tick the clock
+        self.clock.add(1)
+
+        index = self.index_transform(position)
+        target = self.sequence[index].operation
+        op_id = self.clock.get()
+
+        # Add the remove operation to the log and update the sequence
+        op = Operation(OpId(self.id, op_id), OperationType.REMOVE, target)
+        self.operations.add(op)
+        op.do(self.sequence)
+
+    def merge(self, other):
+        """
+        Merges another Sequence with this one.
+        """
+        if not isinstance(other, Sequence):
+            raise ValueError("Incompatible CRDT for merge(), expected Sequence")
+        
+        # Sync the local clock with the remote clock
+        self.clock = self.clock.merge(other.clock)
+
+        # Get the new operations to be applied
+        patch_ops = other.operations.get().difference(self.operations.get())
```

Comment:
We don't have to apply all the operations here, just the ones we don't know about.